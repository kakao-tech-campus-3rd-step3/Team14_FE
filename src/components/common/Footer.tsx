import { useState } from 'react';
import Home from '@/components/icon/HomeIcon';
import Profile from '@/components/icon/ProfileIcon';
import Button from '@/components/common/Button';
import Search from '@/components/icon/SearchIcon';
import Pick from '@/components/icon/PickIcon';

type FooterSelected = 'home' | 'search' | 'pick' | 'my' | 'none';

const navigationItems = [
  { id: 'home', icon: Home, label: '홈' },
  { id: 'search', icon: Search, label: '검색' },
  { id: 'pick', icon: Pick, label: 'AI 추천' },
  { id: 'my', icon: Profile, label: 'My' },
] as const;

const Footer = ({ initialSelected = 'none' }: { initialSelected?: FooterSelected }) => {
  const [selected, setSelected] = useState<FooterSelected>(initialSelected);

  const containerClasses = 'w-full mx-auto flex flex-col items-center fixed bottom-0 left-0 z-999';
  const baseClasses =
    'w-full max-w-[480px] h-15 border-t border-gray-300 flex items-center bg-white';
  const buttonClasses = 'flex items-center justify-center flex-1 flex-col text-xs pt-2 pb-3 px-3';
  const iconClasses = 'size-6';

  return (
    <div className={containerClasses}>
      <div className={baseClasses}>
        {navigationItems.map(({ id, icon: IconComponent, label }) => (
          <Button key={id} className={buttonClasses} variant="icon" onClick={() => setSelected(id)}>
            <IconComponent
              className={iconClasses}
              variant={selected === id ? 'solid' : 'outline'}
            />
            <p className={selected === id ? 'font-bold' : 'font-normal'}>{label}</p>
          </Button>
        ))}
      </div>
    </div>
  );
};

export default Footer;
