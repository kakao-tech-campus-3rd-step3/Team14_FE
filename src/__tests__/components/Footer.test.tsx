import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Footer from '@/components/common/Footer';

describe('Footer 컴포넌트', () => {
  describe('기본 렌더링', () => {
    test('Footer가 렌더링된다', () => {
      // Given: Footer 컴포넌트가 주어졌을 때
      // When: Footer 컴포넌트를 렌더링하면
      const { container } = render(
        <MemoryRouter>
          <Footer />
        </MemoryRouter>,
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('initialSelected 테스트', () => {
    test.each([
      ['home', '홈'],
      ['search', '검색'],
      ['pick', 'AI 추천'],
      ['my', 'My'],
    ] as const)('initialSelected가 %s일 때 해당 버튼이 선택된다', (selected, buttonText) => {
      // Given: 특정 initialSelected가 주어졌을 때
      // When: 해당 initialSelected로 Footer 컴포넌트를 렌더링하면
      render(
        <MemoryRouter>
          <Footer initialSelected={selected} />
        </MemoryRouter>,
      );

      // Then: 해당 버튼이 선택된 상태여야 한다
      const button = screen.getByText(buttonText).closest('button');
      expect(button).toBeInTheDocument();

      // 선택된 버튼의 텍스트가 볼드 처리되어야 한다
      const textElement = screen.getByText(buttonText);
      expect(textElement).toHaveClass('font-bold');
    });

    test('선택되지 않은 버튼들은 normal 폰트를 가진다', () => {
      // Given: home이 선택된 Footer가 주어졌을 때
      // When: initialSelected="home"으로 Footer 컴포넌트를 렌더링하면
      render(
        <MemoryRouter>
          <Footer initialSelected="home" />
        </MemoryRouter>,
      );

      // Then: 선택되지 않은 버튼들은 font-normal 클래스를 가져야 한다
      expect(screen.getByText('검색')).toHaveClass('font-normal');
      expect(screen.getByText('AI 추천')).toHaveClass('font-normal');
      expect(screen.getByText('My')).toHaveClass('font-normal');
    });
  });

  describe('상호작용 테스트', () => {
    test('버튼 클릭 시 폰트가 볼드로 변경된다', () => {
      // Given: Footer 컴포넌트가 렌더링되어 있을 때
      render(
        <MemoryRouter>
          <Footer />
        </MemoryRouter>,
      );

      // When: 홈 버튼을 클릭하면
      const homeButton = screen.getByText('홈');
      fireEvent.click(homeButton.closest('button') as HTMLElement);

      // Then: 홈 버튼의 텍스트가 볼드로 변경되어야 한다
      expect(homeButton).toHaveClass('font-bold');
    });

    test('다른 버튼 클릭 시 이전 선택이 해제되고 새 버튼이 선택된다', () => {
      // Given: 홈이 선택된 상태의 Footer 컴포넌트가 렌더링되어 있을 때
      render(
        <MemoryRouter>
          <Footer initialSelected="home" />
        </MemoryRouter>,
      );

      // When: 검색 버튼을 클릭하면
      const searchButton = screen.getByText('검색');
      const homeButton = screen.getByText('홈');
      fireEvent.click(searchButton.closest('button') as HTMLElement);

      // Then: 검색 버튼은 볼드가 되고 홈 버튼은 normal이 되어야 한다
      expect(searchButton).toHaveClass('font-bold');
      expect(homeButton).toHaveClass('font-normal');
    });

    test('여러 버튼을 순차적으로 클릭할 수 있다', () => {
      // Given: Footer 컴포넌트가 렌더링되어 있을 때
      render(
        <MemoryRouter>
          <Footer />
        </MemoryRouter>,
      );

      // When: 각 버튼을 순차적으로 클릭하면
      const buttonLabels = ['홈', 'AI 추천', 'My'];

      buttonLabels.forEach((label) => {
        const button = screen.getByText(label);
        fireEvent.click(button.closest('button') as HTMLElement);

        // Then: 클릭된 버튼의 텍스트가 볼드로 변경되어야 한다
        expect(button).toHaveClass('font-bold');
      });
    });
  });
});
