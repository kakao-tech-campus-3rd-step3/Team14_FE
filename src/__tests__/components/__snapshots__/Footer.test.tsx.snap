// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer 컴포넌트 > 기본 렌더링 > Footer가 렌더링된다 1`] = `
<div>
  <div
    class="w-full mx-auto flex flex-col items-center fixed bottom-0 left-0 z-999"
  >
    <div
      class="w-full max-w-[480px] h-15 border-t border-gray-300 flex items-center bg-white"
    >
      <button
        class="cursor-pointer flex items-center justify-center w-fit bg-transparent text-gray-900 text-md rounded-xl px-5 py-3 flex items-center justify-center flex-1 flex-col text-xs pt-2 pb-3 px-3"
        type="button"
      >
        <svg
          class="size-6"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3.99999 10L12 3L20 10L20 20H15V16C15 15.2044 14.6839 14.4413 14.1213 13.8787C13.5587 13.3161 12.7956 13 12 13C11.2043 13 10.4413 13.3161 9.87868 13.8787C9.31607 14.4413 9 15.2043 9 16V20H4L3.99999 10Z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <p
          class="font-normal"
        >
          홈
        </p>
      </button>
      <button
        class="cursor-pointer flex items-center justify-center w-fit bg-transparent text-gray-900 text-md rounded-xl px-5 py-3 flex items-center justify-center flex-1 flex-col text-xs pt-2 pb-3 px-3"
        type="button"
      >
        <svg
          class="size-6"
          fill="none"
          height="24"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.7955 15.8111L21 21M18 10.5C18 14.6421 14.6421 18 10.5 18C6.35786 18 3 14.6421 3 10.5C3 6.35786 6.35786 3 10.5 3C14.6421 3 18 6.35786 18 10.5Z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <p
          class="font-normal"
        >
          검색
        </p>
      </button>
      <button
        class="cursor-pointer flex items-center justify-center w-fit bg-transparent text-gray-900 text-md rounded-xl px-5 py-3 flex items-center justify-center flex-1 flex-col text-xs pt-2 pb-3 px-3"
        type="button"
      >
        <svg
          class="size-6"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-width="4"
          viewBox="0 0 64 64"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="34.52"
            cy="11.43"
            r="5.82"
          />
          <circle
            cx="53.63"
            cy="31.6"
            r="5.82"
          />
          <circle
            cx="34.52"
            cy="50.57"
            r="5.82"
          />
          <circle
            cx="15.16"
            cy="42.03"
            r="5.82"
          />
          <circle
            cx="15.16"
            cy="19.27"
            r="5.82"
          />
          <circle
            cx="34.51"
            cy="29.27"
            r="4.7"
          />
          <line
            x1="20.17"
            x2="28.9"
            y1="16.3"
            y2="12.93"
          />
          <line
            x1="38.6"
            x2="49.48"
            y1="15.59"
            y2="27.52"
          />
          <line
            x1="50.07"
            x2="38.67"
            y1="36.2"
            y2="46.49"
          />
          <line
            x1="18.36"
            x2="30.91"
            y1="24.13"
            y2="46.01"
          />
          <line
            x1="20.31"
            x2="28.7"
            y1="44.74"
            y2="48.63"
          />
          <line
            x1="17.34"
            x2="31.37"
            y1="36.63"
            y2="16.32"
          />
          <line
            x1="20.52"
            x2="30.34"
            y1="21.55"
            y2="27.1"
          />
          <line
            x1="39.22"
            x2="47.81"
            y1="29.8"
            y2="30.45"
          />
          <line
            x1="34.51"
            x2="34.52"
            y1="33.98"
            y2="44.74"
          />
        </svg>
        <p
          class="font-normal"
        >
          AI 추천
        </p>
      </button>
      <button
        class="cursor-pointer flex items-center justify-center w-fit bg-transparent text-gray-900 text-md rounded-xl px-5 py-3 flex items-center justify-center flex-1 flex-col text-xs pt-2 pb-3 px-3"
        type="button"
      >
        <svg
          class="size-6"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 21C5 17.134 8.13401 14 12 14C15.866 14 19 17.134 19 21M16 7C16 9.20914 14.2091 11 12 11C9.79086 11 8 9.20914 8 7C8 4.79086 9.79086 3 12 3C14.2091 3 16 4.79086 16 7Z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <p
          class="font-normal"
        >
          My
        </p>
      </button>
    </div>
  </div>
</div>
`;
